core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,292 @@
+    String tableName = this.config.tableName;
+    processedFileTrackingTable = context.getDataset(tableName);
+    if (processedFileTrackingTable != null) {
+      List<String> processedFiles = new ArrayList<String>();
+      Calendar cal = Calendar.getInstance();
+      cal.add(Calendar.DATE, -Integer.valueOf(config.tableExpiryPeriod));
+      Date expiryDate = cal.getTime();
+
+      CloseableIterator<KeyValue<byte[], byte[]>> iterator = processedFileTrackingTable.scan(null, null);
+      while (iterator.hasNext()) {
+        KeyValue<byte[], byte[]> keyValue = iterator.next();
+        //delete record before expiry time period
+        Long time = new Long(new String(keyValue.getValue(), Charsets.UTF_8));
+        Date processedDate = new Date(time);
+        if (processedDate.before(expiryDate)) {
+          processedFileTrackingTable.delete(keyValue.getKey());
+        } else {
+          processedFiles.add(new String(keyValue.getKey(), Charsets.UTF_8));
+        }
+      }
+      //TODO - remove temp file name setting in config after proper solution.
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE, tableName);
+      //For reprocessing not required, set processed file name to configuration.
+      //File name use by BatchXMLFileFilter to filter already processed files.
+      if (config.reprocessingRequired.equalsIgnoreCase("NO")) {
+        conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_FILES,
+                 GSON.toJson(processedFiles, ARRAYLIST_PREPROCESSED_FILES));
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    Map<String, String> xmlRecord = (Map<String, String>) input.getValue();
+    Set<String> keySet = xmlRecord.keySet();
+    Iterator<String>  itr = keySet.iterator();
+    String fileName = itr.next();
+    String record = xmlRecord.get(fileName);
+
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_XML_SCHEMA)
+      .set("offset", input.getKey().get())
+      .set("filename", fileName)
+      .set("record", record)
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    super.onRunFinish(succeeded, context);
+    try {
+      //TODO - remove temp file usage to read processed file data after proper solution implementation.
+      //read file tracking information updated by XMLRecordReader and put it in dataset.
+      File file = new File(this.config.tableName);
+      BufferedReader reader = new BufferedReader(new FileReader(file));
+      String line = null;
+      while ((line = reader.readLine()) != null) {
+        long processingTime = new Date().getTime();
+        processedFileTrackingTable.write(line.getBytes(Charsets.UTF_8),
+                                         String.valueOf(processingTime).getBytes(Charsets.UTF_8));
```

Comment:
Instead of doing long-->string-->bytes, just do:
Bytes.toBytes(long l);

Just make sure you do `Bytes.toLong()`, wherever you're reading this value, to be consistent.